File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int helper(int idx, vector<int>& prices, int buy, vector<vector<int>>& dp) {
        if (idx == prices.size()) return 0;

        if (dp[idx][buy] != -1) return dp[idx][buy];

        int profit;
        if (buy == 1) {
            profit = max(
                -prices[idx] + helper(idx + 1, prices, 0, dp),
                helper(idx + 1, prices, 1, dp)
            );
        } else {
            profit = max(
                prices[idx] + helper(idx + 1, prices, 1, dp),
                helper(idx + 1, prices, 0, dp)
            );
        }

        return dp[idx][buy] = profit;
    }

    int maxProfit(vector<int>& prices) {
        vector<vector<int>> dp(prices.size(), vector<int>(2, -1));
        return helper(0, prices, 1, dp);
    }
};

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {

    int help(int idx , int[] prices , int buy, int[][]dp){
      if(idx == prices.length) return 0;
      if(dp[idx][buy] != -1) return dp[idx][buy];
      if(buy == 1){
          return dp[idx][buy] = Math.max(-prices[idx] + help(idx+1, prices, 0,dp), help(idx+1,prices,1,dp));
      }

      return dp[idx][buy] = Math.max(prices[idx] + help(idx+1, prices, 1,dp), help(idx+1,prices,0,dp));

    }
    public int maxProfit(int[] prices) {
        
        int[][] dp = new int[prices.length][2];

        for(int[] a: dp){
           Arrays.fill(a,-1);
        }

        return help(0,prices,1,dp);
    }
}